File: packages/python/src/opensms/resources/wallet.py
"""The ``wallet`` resource."""

from __future__ import annotations

from typing import Any, List, Optional

from ..models import LedgerEntry, Topup, WalletBalance
from ._base import Resource


def _data(value: Any) -> List[Any]:
    if isinstance(value, dict):
        return list(value.get("data") or [])
    return value or []


class Wallet(Resource):
    """Balances, ledger and top-ups. Accessed as ``client.wallet``."""

    def balances(self) -> List[WalletBalance]:
        """Wallet balances (decimal strings) for the key's environment."""
        return _data(self._t.request("GET", "/v1/wallet"))

    def ledger(self, *, limit: Optional[int] = None, before: Optional[int] = None) -> List[LedgerEntry]:
        """Ledger entries, newest first. Not cursor based: pass the smallest
        ``id`` seen as ``before`` to page, and stop when fewer than ``limit``
        rows come back."""
        return _data(self._t.request("GET", "/v1/wallet/ledger", query={"limit": limit, "before": before}))

    def create_topup(
        self,
        *,
        amount: str,
        currency: str,
        channel: str,
        email: str,
        idempotency_key: Optional[str] = None,
    ) -> Topup:
        """Start a top-up (``channel`` ``card``, ``mobile_money`` or
        ``bank_transfer``). Live keys only. ``amount`` is a decimal string."""
        body = {"amount": str(amount), "currency": currency, "channel": channel, "email": email}
        return self._t.request(
            "POST", "/v1/wallet/topups", json_body=body, idempotent=True, idempotency_key=idempotency_key
        )

File: packages/ruby/lib/opensms/resources/wallet.rb
# frozen_string_literal: true

require_relative "base"

module Opensms
  module Resources
    # The +wallet+ resource: balances, ledger and top-ups. Accessed as
    # +client.wallet+.
    class Wallet < Base
      # GET /v1/wallet -> the +data+ array of balances.
      # @return [Array<Hash>] [{ id:, currency:, balance:, reserved:, environment: }]
      def balances
        unwrap(http_get("/v1/wallet"))
      end

      # Ledger entries, newest first. Pages with +before+ (the smallest id
      # already seen) instead of cursors; stop when fewer than +limit+ rows
      # come back. GET /v1/wallet/ledger.
      #
      # @param params [Hash] :limit (1..200), :before
      # @return [Array<Hash>] LedgerEntry list
      def ledger(params = nil, **kwargs)
        unwrap(http_get("/v1/wallet/ledger", Models.build(merge(params, kwargs), %i[limit before])))
      end

      # Start a payment-provider top-up (live keys only).
      # POST /v1/wallet/topups -> 201 { id:, reference:, authorization_url:, ... }.
      #
      # @param params [Hash] :amount (decimal string), :currency, :channel
      #   (card|mobile_money|bank_transfer), :email
      def create_topup(params = nil, idempotency_key: nil, **kwargs)
        fields = %i[amount currency channel email]
        body = Models.build(merge(params, kwargs), fields, required: fields)
        @transport.request(:post, "/v1/wallet/topups", body: body, idempotency_key: idem(idempotency_key))
      end

      private

      def unwrap(body)
        body.is_a?(Hash) ? (body[:data] || []) : body
      end
    end
  end
end
